### DISCUSSION TITLE
Contextual help isn't always very helpful (command descriptions can be too abstract)

### DISCUSSION THREAD

**Discussion by axelsimon** [neomutt/neomutt#4006] [Development]
## The issue

I had started an email in a graphical client (K9, on Android) which generated an html message. When I went to continue the email in neomutt, I was able to edit the text and finish my email. However, I was left with 2 attachments to my email: my complete message, in plain text, and my original draft, in html. I wanted to delete / remove the unnecessary attachment (the html file) but couldn't remember how to do so. I pressed '?' to call the help menu and typed '/' to search for "attachment", as I wanted to do something with attachment. I ended up spending minutes looking at all the highlighted options, to no avail of course given the desired option is:
`D              –                      detach-file                  –                   delete the current entry`

I have to say, it's not the first time I can't find what I want in the help, even if the answer is simple in hindsight. I had a quick look at the code, it seems to be because the same description of an operation is used even when that operation is called in different contexts:
`"detach-file"` calls `OP_DELETE` but `OP_DELETE` is always described as `delete the current entry`, even if that might really not be most helpful way of putting it, depending on the context.

I thought it was worth sharing this issue, hope it helps.
Thanks.

## NeoMutt Version

```
NeoMutt 20180716
Copyright (C) 1996-2016 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 4.17.12-200.fc28.x86_64 (x86_64)
ncurses: ncurses 6.1.20180224 (compiled with 6.1.20180224)
libidn: 1.34 (compiled with 1.34)
hcache backends: tokyocabinet, lmdb

Compiler:
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --enable-libmpx --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 8.1.1 20180502 (Red Hat 8.1.1-1) (GCC) 

Configure options: --sysconfdir=/etc SENDMAIL=/usr/sbin/sendmail ISPELL=/usr/bin/hunspell --notmuch --tokyocabinet --lmdb --gnutls --sasl --gss --gpgme

Compilation CFLAGS: -g -O2 -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include  -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens +getaddrinfo 
  +gnutls +gpgme +gss +hcache -homespool +idn -locales_hack -lua +meta 
  -mixmaster +nls +notmuch -openssl +pgp +sasl +smime +start_color 
  +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

```

## Extra Info

* Operating System and its version: Fedora F28
* Were you using multiple copies of NeoMutt at once? no
* Were you using 'screen' or 'tmux'? yes, tmux
* Is your email local (maildir) or remote (IMAP)? IMAP

**Comment by flatcap** [comment]
Thanks, @axelsimon, that's a very good point.
My first thought was, "Argh! How are we going to do this?", but it's probably quite simple.

In the code, we reuse, e.g. `OP_DELETE` to mean one of
- `delete-entry`
- `delete-message`
- `detach-file`

but as you noticed `OP_DELETE` is tied to **one** specific help message.
Splitting all these cases would mean we could have help that was more relevant.

**Comment by axelsimon** [comment]
Hi @flatcap, thanks for the positive answer!
Good to hear it might not be as hard as I thought either.

**Comment by axelsimon** [comment]
Here are ~~two~~ one more example of commands that are hard to find, when searching the help:
- ~~edit as new / send again /~~ resend : a command to reopen an already sent email and (possibly) edit it and send it, likely to a different recipient. ~~When viewing the message list, calling up the help and searching for "edit", "new" or "again" doesn't bring up anything helpful.~~
It can be found in the help with the simple term "resend":
`<Esc>e        resend-message           use the current message as a template for a new one`
Well I've learnt something today.
Nevertheless, maybe the help screen's wording could be perfected, so it can catch different people's wording better: coming from Thunderbird I was really looking for "edit" or "again", which are the two fundamental ideas behind this action, I think.
- star a message : one of the simplest ways of marking email for further review, used in Thunderbird, K9 mail, Gmail, etc.. When viewing the message list, calling up the help and searching for "star" or "mark" doesn't bring up anything helpful. Maybe it's called something else, but I couldn't find it :)

**Comment by mschilli87** [comment]
@axelsimon:
> * star a message : one of the simplest ways of marking email for further review, used in Thunderbird, K9 mail, Gmail, etc.. When viewing the message list, calling up the help and searching for "star" or "mark" doesn't bring up anything helpful. Maybe it's called something else, but I couldn't find it :)

Maybe you are looking for a way to '*flag* a message'? :wink:
You could even *tag* several messages by matching a *pattern* and then *flag* all of them at once. :slightly_smiling_face:

**Comment by axelsimon** [comment]
> @axelsimon:
> 
> > * star a message : one of the simplest ways of marking email for further review, used in Thunderbird, K9 mail, Gmail, etc.. When viewing the message list, calling up the help and searching for "star" or "mark" doesn't bring up anything helpful. Maybe it's called something else, but I couldn't find it :)
> 
> Maybe you are looking for a way to '_flag_ a message'? wink
> You could even _tag_ several messages by matching a _pattern_ and then _flag_ all of them at once. slightly_smiling_face

Thanks, that's helpful :slightly_smiling_face:  I think (I hope!) we can agree on the fact that the wording could be tweaked to include the words a lot of people use to describe these actions :wink:

**Comment by mschilli87** [comment]
>I think (I hope!) we can agree on the fact that the wording could be tweaked to include the words a lot of people use to describe these actions

Agreed. I found most of the things by calling up the help by <kbd>?</kbd> and then scrolling via <kbd>Page Down</kbd> skimming through the text (*i.e.* '`grep`ing' by eye). :wink:

**Comment by flatcap** [comment]
There are 13 menus defined in [functions.h](https://github.com/neomutt/neomutt/blob/master/functions.h)
- OpGeneric
- OpMain
- OpPager
  ...

These link a string "save-message" to a constant `OP_SAVE`.
[opcodes.h](https://github.com/neomutt/neomutt/blob/master/opcodes.h) defines these constants and associates help text with each.

I'd really appreciate it, if someone had the time to have a look through the list and provide some alternate text.

Excluding the overlap between the Index and Pager menus, there are about 50 duplicates
(most won't need changing).

```
OP_BOUNCE_MESSAGE
OP_CATCHUP
OP_CHECK_STATS
OP_CHECK_TRADITIONAL
OP_COMPOSE_TO_SENDER
OP_CREATE_ALIAS
OP_DELETE
OP_DISPLAY_HEADERS
OP_EDITOR_MAILBOX_CYCLE
OP_EDIT_TYPE
OP_ENTER_COMMAND
OP_EXIT
OP_EXTRACT_KEYS
OP_FOLLOWUP
OP_FORGET_PASSPHRASE
OP_FORWARD_MESSAGE
OP_FORWARD_TO_GROUP
OP_GENERIC_SELECT_ENTRY
OP_GROUP_REPLY
OP_HALF_DOWN
OP_HALF_UP
OP_HELP
OP_JUMP
OP_LIST_REPLY
OP_MAIL
OP_MAILBOX_LIST
OP_MIX_USE
OP_NEXT_ENTRY
OP_NEXT_LINE
OP_NEXT_PAGE
OP_PIPE
OP_PREV_ENTRY
OP_PREV_LINE
OP_PREV_PAGE
OP_PRINT
OP_QUERY
OP_REDRAW
OP_REPLY
OP_RESEND
OP_SAVE
OP_SEARCH
OP_SEARCH_NEXT
OP_SEARCH_OPPOSITE
OP_SEARCH_REVERSE
OP_SHELL_ESCAPE
OP_SORT
OP_SORT_REVERSE
OP_TAG
OP_UNDELETE
OP_VERIFY_KEY
OP_VIEW_ATTACH
OP_VIEW_ID
OP_WHAT_KEY
```

**Comment by axelsimon** [comment]
Here is some more stuff that's hard to find:
- recalling drafts: the word "draft" doesn't appear once in the help You have to search the help menu for "postponed". I don't think I know anyone who refers to draft emails as "postponed messages".

**Comment by flatcap** [comment]
> the word "draft" doesn't appear once in the help

hehe, thanks.  Noted.

One of my 'Big Plans' is to embed the help **in** NeoMutt as a fake mailbox.
We'd break up the docs into small pieces and each would have metadata, like 'keywords'.
Searching priority would match Titles, Keywords and then Body text.

**Comment by axelsimon** [comment]
That actually sounds nice!

It should make it possible to put the most sought after info first, and the less often used commands after (although i'm sure i'm the only one who doesn't use "edit attachment content type" everyday when everyone else does :wink:).

And it could even serve as a built-in tutorial on how to use NeoMutt. The first email tells you to press `<space>` to read the next screen. Then to press `<tab>`to jump to the next email, which itself gives you more info as to what to do next, etc.

**Comment by axelsimon** [comment]
> * edit as new / send again / resend : a command to reopen an already sent email and (possibly) edit it and send it, likely to a different recipient. When viewing the message list, calling up the help and searching for "edit", "new" or "again" doesn't bring up anything helpful.
>  It can be found in the help with the simple term "resend":
>  `<Esc>e        resend-message           use the current message as a template for a new one`

I was thinking about this again. Could we simply add change the text to `use the current message as a template for a new one (ie: edit as new)`?

Would editing [opcodes.h](https://github.com/neomutt/neomutt/blob/master/opcodes.h) be enough to do that?

**Comment by flatcap** [comment]
> put the most sought after info first

Hmm... I hadn't actually thought about _that_.
Adding a weighting to the information would be easy enough.

> it could even serve as a built-in tutorial

Yes, that'd be nice.  Something like the vim-tutor.
Each page would tell you:
- Where you are
- What you can do next (including aborting)
- (How you got there)

> Could we simply add change the text ... editing opcodes.h

I'm happy to accept improvements for the help text...

however distinguishing between "Delete email", "Delete alias", etc will mean creating new constants and adding them to the Index/Alias/etc menus.

I don't have the time, myself, but I'm happy to help someone work through it.
